### PR TITLE
Support credential update

### LINKED ISSUE #65: Allow to update credential
Instead of overwriting the credential, allow to update it with the new values of the provided fields.

Connected: 
- #55 
- https://github.com/Nitrokey/pynitrokey/issues/402

### PR DESCRIPTION
**Pull request by szszszsz**
Add Initial support for the credential update
Makes RenameCredential command obsolete

Fixes #65 

pynitrokey support provided in https://github.com/Nitrokey/pynitrokey/pull/425

### CODE REVIEW

File: src/command.rs [hunk 1]
```diff
@@ -518,6 +520,94 @@ pub struct Credential<'l> {
...
+        };
+
+        // TODO This would be better modeled with Option, than empty slice,
+        //  to allow fields removal by setting the request fields to empty slice.
+        //  Currently only setting is possible. A workaround for now is to
+        //  set the value to remove to " ".
+        let mut pws = PasswordSafeData {
+            login: &[],
+            password: &[],
+            metadata: &[],
```
Comment (sosthene-nitrokey) [inline src/command.rs:561]: Why not do it now with an option?
I feel like it's better to do it the right way from the start.

Also with an option `None` could mean "unchanged"  (as I would expect most updates to only change one field).

Comment (sosthene-nitrokey) [reply src/command.rs:561]: I would actually do it with:

- `None` means unchanged
- `Some([])` means delete (i think we may want to reject empty slices also at registration
- `Some([..])` means update this part.

This feels more intuitive (to me), and is consistent with how deletion works in PIV for example.

Comment (szszszsz) [reply src/command.rs:561]: Yup, this sounds good. I've intended to this this way (as in comment).
This looked like a bigger refactor (as you've mentioned, involves register command), hence I wanted to extract this to a separate task to keep it clear and focused.

Comment (sosthene-nitrokey) [reply src/command.rs:561]: I don't think it is big enough to justify be done in another PR, but why not.

I still think we need to hold merging https://github.com/Nitrokey/pynitrokey/pull/425#pullrequestreview-1576442445 in that case before the refactor is done.

Comment (szszszsz) [reply src/command.rs]: Updated as mentioned. Resolving.

File: src/command.rs [hunk 2]
```diff
@@ -518,6 +518,86 @@ pub struct Credential<'l> {
     pub password_safe: Option<PasswordSafeData<'l>>,
 }
 
+#[derive(Clone, Copy, Eq, PartialEq, Debug, Default)]
+pub struct CredentialUpdate<'l> {
```
Comment (robin-nitrokey) [inline src/command.rs:522]: nit: why not `UpdateCredential`?  For the other commands, the structs and the `Command` variants use the same name.

File: src/credential.rs [hunk 3]
```diff
@@ -201,11 +206,30 @@ impl CredentialFlat {
...
+        if let Some(p) = update_req.properties {
+            self.touch_required = p.touch_required();
+        }
+        if let Some(pws) = update_req.password_safe {
+            self.login = Self::get_bytes_if_not_empty_or_none(pws.login)
+                .map_err(|_| Status::UnspecifiedNonpersistentExecutionError)?;
+            self.password = Self::get_bytes_if_not_empty_or_none(pws.password)
+                .map_err(|_| Status::UnspecifiedNonpersistentExecutionError)?;
+            self.metadata = Self::get_bytes_if_not_empty_or_none(pws.metadata)
+                .map_err(|_| Status::UnspecifiedNonpersistentExecutionError)?;
```
Comment (robin-nitrokey) [inline src/credential.rs:231]: Shouldn’t we ignore `None` values here (instead of returning an error)?

File: src/command.rs (hunk 2 above)
Comment (szszszsz) [reply src/command.rs:522]: Yup, I saw that too, but too late. I can make a quickfix later.

File: src/credential.rs (hunk 3 above)
Comment (szszszsz) [reply src/credential.rs:231]: Error is only on conversion issue, not if None is encountered. `.map()` would be ideal here, but I had problem with dealing with the said error in the `Fn`.

File: src/command.rs (hunk 2 above)
Comment (szszszsz) [reply src/command.rs]: Done

**Comment by szszszsz** [comment]
All raised concerns are solved. Merging.

**Pull request merged**